### DISCUSSION TITLE
Order of query match traversal?

### DISCUSSION THREAD

**Discussion by HiPhish** [tree-sitter/tree-sitter#2067] [Q&A] [answered]
Hello,

I am a Neovim user and I am trying to make rainbow delimiters work, i.e. use Tree-sitter to apply alternating highlighting to delimiters like parentheses or HTML tags. Everything works so far except for one thing: I cannot figure out the nesting level of a node relative to other delimiter nodes.

Getting the absolute nesting level is easy, just keep going through parent nodes and keep counting until we hit the root. However, that is not what I need. I need to skip over the nodes which are not some sort of delimited container. So here is my idea: if (and that's a hard if) there is a guarantee about the order of tree traversal when matching a query against the root node, then I can determine the nesting level of all matched nodes.

-----

Please allow me to illustrate an example. Here is a snippet of HTML code:

```html
<a href="https://example.com">
    Example<br/>link
</a>
```

The entire block is a node of type element and I want to highlight the opening and closing tags. The query is

```
(element
   ((start_tag) @opening
    (end_tag) @closing)) @container
```

Note the `@container` capture group. I want to know the level of the `@container` node relative to other `@container` nodes.  HTML is simple because there is only one pattern for the entire language, but other languages are more complicated. For example, in C I might have a parenthesized expression inside an if-block inside a for-loop inside a function definition. All of these will be matched by different patterns, but each pattern will have an outer-most `@container` capture. Here are some of the patterns I have so far for C:

```
(parenthesized_expression
  (("(" @opening)
   (")" @closing))) @container

(compound_statement
  (("{" @opening)
   ("}" @closing))) @container

(initializer_list
  (("{" @opening)
   ("}" @closing))) @container

(subscript_expression
  (("[" @opening)
   ("]" @closing))) @container
```

I know I can match a node, such as the document root, against a query and use a query cursor to iterate through the matches. I know I can use `ts_query_cursor_exec` to perform the match and `ts_query_cursor_next_match` to iterate through the matches, but the [documentation](https://tree-sitter.github.io/tree-sitter/using-parsers#the-query-api) does not say anything about the order. If the order is guaranteed to be depth-first it would be very easy for me to achieve my goal by keeping a stack of nodes.

**Comment by HiPhish** [comment]
I have looked at the code and the documentation comment for `ts_query_cursor_new` states the following:

> Repeatedly call `ts_query_cursor_next_match` to iterate over all of the *matches* in the order that they were found.

However, there is no mention of what this order is. It seems reasonably to me that the order would be the same order as the document, i.e. depth-first, but there is no certainty promised.

**Comment by ahelwer** [comment]
Couldn't you determine nest order by looking at the byte range of the container? Nested container byte ranges should be fully contained within a larger container, right?

**Reply by HiPhish** [reply]
That would required keeping track of all nodes ever encountered. Take the following tree as an example:

```
A
├──B
├──C
│  └─D
├──E
│  └─F
└──G
   └─H
     └─I
```

If I want to know the nesting level of node `I` I need to know nodes `A`, `G` and `H`. However, without any guarantee of order I would only be certain that I know all nodes after I have iterated through the entire tree. I would need to build up a "mirror tree" of all matches as I move along the document.

However, if tree traversal is depth-first I can get by with just a stack using a simple algorithm:

1) Start with an empty stack and the first node
2) Is the current node contained in the top-most node of the stack? (if the stack is empty then the answer is always "no")
3) If no, then keep popping nodes off the stack until the answer is yes or the stack is empty. Push the current node onto the stack. The nesting levels of the node is the size of the stack
4) If yes, then push the node on the stack. The nesting levels of the node is the size of the stack
5) Set the next node as the current node, GOTO 2)

If there is no next node the algorithm terminates. I want to keep things as tight as possible since the code will potentially execute on every cursor movement.

**Reply by ahelwer** [reply]
I don't understand why you're traversing the tree instead of just doing queries? I would:
1. Do a query for all parentheses nodes
2. Iterate over all captures, adding them to a data structure that contains both the start/end paren byte locations as well as the overall byte range in the file encompassed by the start/end tokens
3. Sort the data structures into an array of stacks based on whether one byte range is contained within another
4. For each element in the array, emit the associated nest color for each element in the stack's start/end paren byte locations

Why does that not work?

**Reply by HiPhish** [reply]
Correct, I do want to iterate over all matches, not the entire document tree. My question is whether there is a guarantee regarding the order of matches returned because I want to avoid having to collect all matches first and then building a tree of all matches.

**Reply by ahelwer** [reply]
In my experience matches come in the order they occur in the file, but not sure how that works for nested matches. I assume it will either be whichever starts first or whichever ends first. Try it and see?

**Comment by maxbrunsfeld** [comment, marked as answer]
The matches are returned in the order that their final node is found, in a depth first walk of the tree.

**Reply by HiPhish** [reply]
Thank you, that solves the issue. It's the visitor pattern, but the match object is only produced on exiting the node. This means that if we have the following tree of matches, the order will be `B`, `D`, `F`, `G`, `E`, `C`, `A`.

```
A
├─B
└─C
  ├─D
  └─E
    ├─F
    └─G
```

That complicates matter quite a bit, but I have an algorithm that works:

0. Start with an empty set of matches `s = {}`
1. For each match `m` do the following
2. Remove all matches whose `container` node is contained in the container node of `m` from `s`, collect them in a set `s_m`
3. Set `s_m` to be the child set of `m`
4. Add `m` to `s`

This algorithm makes use of the fact that we never skip over ancestor matches. For example, above `E` always follows `F` and `G`, we don't skip over `E` before we get to `C` or `A`. The downside is that we have to build up the entire matches tree for the document before we can know the level of nesting because we never know how deep we are until we are through all matches. If the matches were produced on entering  the node we could immediately know the nesting because the first match would always have a level of 1.

Here is an example of rainbow delimiters in Common Lisp in Neovim:

![Screenshot_20230205_141340](https://user-images.githubusercontent.com/4954650/216821912-d6eade9e-921c-4620-93d9-5cfe38b3b336.png)

Parentheses at the same level have the same colouring.
